packages/portfolio-contract/src/schedule-order.ts

```diff
@@ -0,0 +1,97 @@
+/**
+ * @file schedule async tasks based on partial order
+ * @see {runJob}
+ */
+
+/** zero based index */
+type Ix = number;
+export type Job = {
+  taskQty: number;
+  /** entries of map from vertex to vertexes it depends on */
+  order: Array<[Ix, Ix[]]>;
+};
+
+const range = (n: number) => Array.from(Array(n).keys());
+
+const ok = {
+  status: 'fulfilled',
+  value: undefined,
+} as PromiseSettledResult<void>;
+harden(ok);
+
+/**
+ * call runTask(ix, ...) for each 0 <= ix < job.taskQty,
+ * only when dependent tasks are finished.
+ */
+export const runJob = async (
+  job: Job,
+  runTask: (ix: Ix, running: number[]) => Promise<void>,
+  trace: (...args: unknown[]) => void,
+): Promise<PromiseSettledResult<void>[]> => {
+  const running = new Map<Ix, Promise<Ix>>();
+
+  const { taskQty } = job;
+  const taskIxs = range(taskQty);
+  const todo = new Set(taskIxs);
+  const results = taskIxs.map(_ => ok);
+
+  /** keys with empty dependencies are omitted */
+  const order: Map<Ix, Set<Ix>> = new Map(
+    job.order.flatMap(([ix, deps]) =>
+      deps.length > 0 ? [[ix, new Set(deps)]] : [],
+    ),
+  );
+  const ready = (ix: Ix) => !order.has(ix);
+
+  const failTaskAndAncestors = (ix: Ix, reason: unknown) => {
+    trace('fail', ix, reason);
+    todo.delete(ix);
+    results[ix] = { status: 'rejected', reason };
+
+    const cascade = Error(`predecessor ${ix} failed`, { cause: reason });
+    for (const [candidate, deps] of order.entries()) {
+      if (deps.has(ix)) {
+        failTaskAndAncestors(candidate, cascade);
+      }
+    }
+  };
+
+  await null;
+
+  while (todo.size > 0) {
+    const runnable = [...todo].filter(v => ready(v) && !running.has(v));
+    // trace('runnable', ...runnable);
+    if (!runnable.length && !running.size) {
+      trace('loop! todo', ...todo, 'running', ...running.keys());
+      throw Error('loop!');
```

Comment:
I'd prefer to see something more explanatory, like:

```suggestion
      throw Error('Job dependency loop would prevent further progress');
```